### DISCUSSION TITLE
Custom logic of updating the Machine status

### DISCUSSION THREAD

**Discussion by pparshin** [kubernetes-sigs/cluster-api#5363] [Q&A] [answered]
Hi there,

I'm using my own version of `kubeadm` to deploy the control plane components as unix services. As a result there are no master nodes in the Kubernetes cluster and `MachineController` reconciles in infinite loop attempting to find the master node and set the Running status to Machine.

I walked through the code and find this [logic](https://github.com/kubernetes-sigs/cluster-api/blob/6d1f69218935b43b4783758c579e96c7072a6803/controllers/machine_controller_noderef.go#L64). What should I do to update the Machine status manually or should I use a custom operator? Should I completely rewrite the Cluster API logic to my own or it is possible to extend it?

**Comment by enxebre** [comment, marked as answer]
Hey @pparshin not sure how the linked reference is relevant, that's only trying to find the counter part nodes for existing machines.
CAPI composability principles enables your use case by separating the control plane from compute for workloads. You can define and experiment with your own controlPlane as far some contract requirements are met via CR ref https://github.com/kubernetes-sigs/cluster-api/blob/main/api/v1beta1/cluster_types.go#L57.

See EKS https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/main/controlplane/eks/api/v1alpha4/awsmanagedcontrolplane_types.go or hypershift as examples https://github.com/openshift/hypershift/blob/main/api/v1alpha1/hosted_controlplane.go

If this does not answer your question please let's open an issue with your particular ask and keep the discussion there. Thanks!
